tests/end_to_end/end_to_end_lib.py

```diff
@@ -257,12 +260,21 @@ def generate_config_and_config_file(config_filename, num_of_master_server,
 def start_master_and_chunk_servers(config_filename, log_directory = None,
         num_of_master_server = 1, num_of_chunk_server = 3, block_size_mb = 64,
         min_free_disk_space_mb = 100, grpc_timeout_s = 10, lease_timeout_s = 60,
-        heartbeat_timeout_s = 30, client_cache_timeout_m = 10):
-    # Generate the config file and config data
-    config_data = generate_config_and_config_file(config_filename, 
-                      num_of_master_server, num_of_chunk_server, block_size_mb,
-                      min_free_disk_space_mb, grpc_timeout_s, lease_timeout_s,
-                      heartbeat_timeout_s, client_cache_timeout_m)
+        heartbeat_timeout_s = 30, client_cache_timeout_m = 10,
+        start_master = True, start_chunk = True):
+    global config_data_global
+    
+    # Generate the config file and config data if this function gets the first 
+    # time
+    if not config_data_global:
```

Comment:
this will make subsequent calls that try to initialize different config, reuse previous configs, which defeats the purpose of passing these extra parameters as function arguments; and if that's the case, we should just remove them from function arguments of `start_master_and_chunk_servers` completely

Reply:
Just to clarify, multiple processes (therefore multiple tests) are in separate address spaces so own their own global variable and do not share anything. 

This change potentially affects a case that one wants to call this function multiple times and start multiple clusters. This seems like a case that is beyond our current scope. 

The purpose of introducing this change is to allow that we first launch say master, then sometime later launch the chunk servers. Because port numbers are generated randomly each time, I need something to keep the state. 

There is certainly room for improvement, but I am kinda running out of time.